runtime/src/mm/cpp/ObjectFactory.cpp

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+#include "ObjectFactory.hpp"
+
+#include "Alignment.hpp"
+#include "Alloc.h"
+#include "GlobalData.hpp"
+#include "Types.h"
+
+using namespace kotlin;
+
+ObjHeader* mm::ObjectFactory::ThreadQueue::CreateObject(const TypeInfo* typeInfo) noexcept {
+    RuntimeAssert(!typeInfo->IsArray(), "Must not be an array");
+    size_t allocSize = typeInfo->instanceSize_;
+    auto& node = producer_.Insert(allocSize);
+    auto* object = static_cast<ObjHeader*>(node.Data());
+    object->typeInfoOrMeta_ = const_cast<TypeInfo*>(typeInfo);
+    return object;
+}
+
+ArrayHeader* mm::ObjectFactory::ThreadQueue::CreateArray(const TypeInfo* typeInfo, uint32_t count) noexcept {
+    RuntimeAssert(typeInfo->IsArray(), "Must be an array");
+    uint32_t arraySize = static_cast<uint32_t>(-typeInfo->instanceSize_) * count;
+    // Note: array body is aligned, but for size computation it is enough to align the sum.
+    size_t allocSize = AlignUp(sizeof(ArrayHeader) + arraySize, kObjectAlignment);
```

Comment:
Is this correct that `instanceSize` already includes a padding needed to properly align each element of the array (like `sizeof` does for C++ classes) so we don't need to think about this padding when computing the `allocSize`?

Reply:
It should, yes. At least this matches what's happening in `legacymm`.